text/0000-dyn-assoc.md

```diff
@@ -0,0 +1,371 @@
+
+Explicitly using one of the vtables would look like this:
+```rust
+trait Foo {
+    const CONST: usize;
+}
+fn foo<T: Foo + ?Sized>(t1: &T, t2: &T) -> usize {
+    t1.CONST           // This line changed!
+}
+struct DefaultFoo;
+impl Foo for DefaultFoo {
+    const CONST: usize = 42;
+}
+struct OtherFoo;
+impl Foo for OtherFoo {
+    const CONST: usize = 87;
+}
+
+fn main() {
+    let tobject1: Box<dyn Foo> = Box::from(DefaultFoo);
+    let tobject2: Box<dyn Foo> = Box::from(OtherFoo);
+
+    foo(tobject1.as_ref(), tobject2.as_ref());
+}
+```
+In this example, `foo` explicitly uses one specific vtable, removing any kind of ambiguity. This does not include any runtime checks, and is generally preferrable.
+
+Types with generic arguments work the exact same way: potentially having multiple trait objects of the same trait but of different vtables either requires an equality bound or method-syntax access.
+
+This is only an issue if the generic type parameter has a `?Sized` bound. Otherwise, it's fine to use the normal `T::assoc` syntax.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The main point of technical discussion around this feature is how would dynamic dispatch work if there's no `&self`. Consider the first example from the [guide-level section][guide-level-explanation]:
+```rust
+trait ThingDoer {
+    fn do_a_thing();
+}
+struct HWThingDoer;
+impl ThingDoer for HWThingDoer {
+    fn do_a_thing() {
+        println!("Did a thing using the GPU at insane speed!");
+    }
+}
+struct DefaultThingDoer;
+impl ThingDoer for DefaultThingDoer {
+    fn do_a_thing() {
+        println!("Did a thing entirely on the CPU.");
+    }
+}
+
+fn main() {
+    let mut thingdoer: Box<dyn ThingDoer> = Box::from(HWThingDoer);
+    thingdoer.do_a_thing();
+    thingdoer = Box::from(DefaultThingDoer);
+    thingdoer.do_a_thing();
+}
+```
+What's going on here, exactly? How does the program find the implementation if we don't have access to `&self`? **We actually do.** Since dynamic dispatch happens in `main` (the calling function), rather than a stub `do_a_thing` "disambiguator" which reads the trait object and redirects execution according to its vtable. Simply put, the entire technical idea of this RFC is *if we have access to the trait object when we're doing dynamic dispatch, why do we pretend that the dynamically dispatched implementation needs `&self` in order to understand what to do when it actually doesn't?*
+
+## Associated constants
+It's not any harder for associated constants, since these can be simply stored in the vtable along with the methods and associated functions. The only reason why this might become an implementation issue is that the vtable lookup code can no longer assume that all elements have the same size alignment. It's of high doubt that this assumption is ever made anywhere inside the compiler, though.
```

Comment:
The compiler may have assumed that the v-table has the same layout as a `[usize; N]` where `N` is the number of elements in the v-table (drop info, and functions).

Reply:
I doubt it ever did. Something like this:
```rust
struct Vtable {
    size: usize,
    methods_and_associated_fns: [*const (), N],
    constants: (TypeOfConstant1, TypeOfConstant2, TypeOfConstantN)
}
```
would still work just fine for that purpose.